File: openoversea/src/main/java/com/bytedance/sdk/open/tiktok/utils/SignatureUtils.java
package com.bytedance.sdk.open.tiktok.utils;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.pm.Signature;
import android.text.TextUtils;

import java.util.ArrayList;
import java.util.List;


public class SignatureUtils {

    /**
     * verify signature
     *
     * @param context
     * @param pkgName
     * @param sign
     * @return
     */
    public static boolean validateSign(Context context, String pkgName, String sign) {
        if (TextUtils.isEmpty(pkgName) || TextUtils.isEmpty(sign) || context == null) {
            return false;
        } else {
            List<String> signList = getMd5Signs(context, pkgName);
            if (signList != null && signList.size() > 0) {
                for (String appSign : signList) {
                    if (sign.equalsIgnoreCase(appSign)) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    /**
     * get app signature
     *
     * @param context
     * @param pkgName
     * @return
     */
    public static List<String> getMd5Signs(Context context, String pkgName) {
        if (TextUtils.isEmpty(pkgName)) {
            return null;
        } else {
            PackageInfo packageInfo;
            try {
                packageInfo = context.getPackageManager().getPackageInfo(pkgName, PackageManager.GET_SIGNATURES);
            } catch (PackageManager.NameNotFoundException var4) {
                return null;
            }
            List<String> signList = new ArrayList<>();
            Signature[] signatures = packageInfo.signatures;
            if (null != signatures) {
                Signature[] var2 = signatures;
                int var3 = signatures.length;
                for (int var4 = 0; var4 < var3; ++var4) {
                    Signature signature = var2[var4];
                    String s = Md5Utils.hexDigest(signature.toByteArray());
                    signList.add(s);
                }
                return signList;
            }
        }
        return null;
    }

    /**
     *
     * Joining signatures in an array into a string
     * @param signs
     * @return
     */
    public static String packageSignature(List<String> signs) {
        if (signs != null && !signs.isEmpty()) {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < signs.size(); i++) {
                if (i != 0) {
                    sb.append(",");
                }
                sb.append(signs.get(i));
            }
            return sb.toString();
        }
        return null;
    }
}

File: openoversea/src/main/java/com/bytedance/sdk/open/tiktok/utils/SignatureUtils.kt
package com.bytedance.sdk.open.tiktok.utils

/*
    Copyright 2022 TikTok Pte. Ltd.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
 */

import android.content.Context
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.text.TextUtils

internal object SignatureUtils {
    // validate tiktok or musically app's signature
    fun validateSign(context: Context?, pkgName: String, sign: String): Boolean {
        if (TextUtils.isEmpty(pkgName) || TextUtils.isEmpty(sign) || context == null) {
            return false
        }
        val signList = getMd5Signs(context, pkgName)
        return signList?.find {
            sign.equals(it, ignoreCase = true)
        } != null
    }

    fun getMd5Signs(context: Context, pkgName: String?): List<String?>? {
        if (pkgName.isNullOrEmpty()) {
            return null
        }
        val packageInfo: PackageInfo = try {
            context.packageManager.getPackageInfo(pkgName!!, PackageManager.GET_SIGNATURES)
        } catch (ex: PackageManager.NameNotFoundException) {
            return null
        }
        val signList: List<String?>? = packageInfo.signatures?.map {
            Md5Utils.hexDigest(it.toByteArray())
        }
        return signList
    }

    fun packageSignature(signs: List<String?>?): String? {
        if (signs != null && !signs.isEmpty()) {
            val sb = StringBuilder()
            for (sign in signs) {
                sb.append(",").append(sign)
            }
            return if (sb.isEmpty()) null else sb.substring(1)
        }
        return null
    }
}
